### PR TITLE
[SYCL] Do not use handler::addAccessorReq()

### PR DESCRIPTION
**Pull request by Alexandr-Konovalov**
This function might be useful only when an accessor is connected with different
CG that was used during a construction of the accessor. Currently it’s helpful 
only in single case of reduction, and in this case the accessor can be connected
to a right CG in the constructor. Modify the use case and remove all calls to 
addAccessorReq().

### CODE REVIEW

File: sycl/source/handler.cpp [hunk 1]
```diff
@@ -2055,6 +2055,8 @@ void handler::SetHostTask(std::function<void(interop_handle)> &&Func) {
   setType(detail::CGType::CodeplayHostTask);
 }
 
+// TODO: This function is not used anymore, remove it in the next
+// ABI-breaking window.
 void handler::addAccessorReq(detail::AccessorImplPtr Accessor) {
```
Comment (uditagarwal97) [inline sycl/source/handler.cpp:2060]: We should put it under the -fpreview-breaking-flag. Something like:
```
#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
void handler::addAccessorReq() {
...
}
#endif
```
We use this macro to track all ABI-breaking changes. So that, during the next ABI-breaking window, we can make the code under `ifdef __INTEL_PREVIEW_BREAKING_CHANGES` default.

File: sycl/include/sycl/reduction.hpp [hunk 2]
```diff
@@ -1014,7 +1014,7 @@ class reduction_impl_algo {
             }
           });
         } else {
-          accessor OutAcc{Out, CGH};
+          accessor OutAcc{Out, CopyHandler};
```
Comment (uditagarwal97) [inline sycl/include/sycl/reduction.hpp:1017]: Sorry, but I don't seem to understand why this change was made? Can you please elaborate?

Comment (Alexandr-Konovalov) [reply sycl/include/sycl/reduction.hpp:1017]: @uditagarwal97 , good point, it might deserve an extended comment in the commit message.

In old behavior, `OutAcc` was connected with `CGH`, but this is not enough for the correct operations, so later same accessor was connect with `CopyHandler` via `addAccessorReq()`. But connection with `CGH` is not required, thus allowing us to connect with `CopyHandler` right away and eliminate `addAccessorReq()` calls.

File: sycl/source/handler.cpp (hunk 1 above)
Comment (Alexandr-Konovalov) [reply sycl/source/handler.cpp:2061]: Done.

**Review by uditagarwal97 — APPROVED** [review]

**Comment by Alexandr-Konovalov** [comment]
@intel/llvm-gatekeepers Could you please look at the PR?

**Comment by steffenlarsen** [comment]
My biggest worry about this patch is that we no longer keep accessors alive in the accessor storage of the command-group handler for many of these paths. Is this safe? Are we confident there are no scenarios that need them alive? I can see the graph copies the storage, @intel/sycl-graphs-reviewers.

**Comment by Bensuo** [comment]
> My biggest worry about this patch is that we no longer keep accessors alive in the accessor storage of the command-group handler for many of these paths. Is this safe? Are we confident there are no scenarios that need them alive? I can see the graph copies the storage, @intel/sycl-graphs-reviewers.

I think this change should be fine for us in general if I'm understanding it correctly. Accessors are by default stored when associated with a handler (either through the ctor taking a handler or `handler::require()`) [via this code](https://github.com/intel/llvm/blob/afa88218f4cc59ae27c1ea4d2b4a0c153a0ece45/sycl/source/handler.cpp#L590). So outside of this niche usecase where the accessors needed to be associated with an additional handler I think this would have just been creating duplicate requirements for calls that took these paths.

We don't currently support reductions in graphs but I don't think this change makes much difference to the challenges of us implementing support for them so it seems fine in that regard also.

**Review by Bensuo — APPROVED** [review]

**Comment by steffenlarsen** [comment]
Thanks, @Bensuo! In that case, I forfeit my concerns. Let's roll with it!
